### DISCUSSION TITLE
What is the difference between ignite's metrics and pytorch's metrics?

### DISCUSSION THREAD

**Discussion by Jingnan-Jia** [pytorch/ignite#1548] [Q&A]
I found that ignite have its own group of metrics. What is the difference with pure pytorch's metrics? Why does ignite not use pytorch's metrics directly?

update:
Here the `pure pytorch's metrics` mean pytorch's `loss` function.

**Comment by vfdev-5** [comment]
@Ordgod which "pure pytorch's metrics" you are talking about ?

**Reply by tarasio-mirror** [reply]
@vfdev-5 I suppose  loss functions? But we can use them with `ignite`'s `Loss` metric. Am I right?

**Reply by vfdev-5** [reply]
Metrics were introduced to ignite almost since its begining. As far as I know, pytorch itself does not provide any metrics to compute. We did not introduced various SOTA loss functions as torch already has a number of built-in loss functions such that we wont overlap with it. `Loss` metric is a wrapper to compute average loss value for entire epoch (and not a single iteration).

**Comment by importstring** [comment]
### For PyTorch:

```python
# Training-time: use loss functions batch-by-batch
for x, y in dataloader:
    y_pred = model(x)
    loss = loss_fn(y_pred, y)      # e.g. nn.CrossEntropyLoss
    loss.backward()                # drives optimization
    optimizer.step()
    optimizer.zero_grad()
```
### For PyTorch-Ignite:

```python
# Evaluation-time: use Ignite metrics over an epoch
from ignite.engine import Engine
from ignite.metrics import Accuracy, Loss

def step_fn(engine, batch):
    x, y = batch
    y_pred = model(x)
    return y_pred, y              # engine.state.output

evaluator = Engine(step_fn)
Accuracy().attach(evaluator, "accuracy")
Loss(loss_fn).attach(evaluator, "avg_loss")  # wraps PyTorch loss_fn

evaluator.run(val_loader)
metrics = evaluator.state.metrics  # epoch-level metrics
```
And these different roles give you finer control over training vs. evaluation: PyTorch’s loss functions are low-level building blocks for optimization on each batch, while Ignite’s metrics are high-level components that accumulate and compute metrics (including average loss) over an entire epoch, integrate with the event engine, and support distributed evaluation out of the box.
